toolchain/lowering/lowering_context.cpp

```diff
@@ -103,15 +103,15 @@ auto LoweringContext::BuildFunctionDefinition(SemanticsFunctionId function_id)
     function_lowering.SetLocal(param_storage, llvm_function->getArg(i));
   }
 
-  // Add the entry block to the worklist.
-  function_lowering.GetBlock(function.body_id);
-
-  while (!function_lowering.block_worklist().empty()) {
-    SemanticsNodeBlockId block = function_lowering.block_worklist().Pop();
-    CARBON_VLOG() << "Lowering " << block << "\n";
-    function_lowering.builder().SetInsertPoint(
-        function_lowering.GetBlock(block));
-    for (const auto& node_id : semantics_ir().GetNodeBlock(block)) {
+  // Lower all blocks.
+  // TODO: Determine the set of reachable blocks, and only lower those ones.
+  for (auto block_id : body_block_ids) {
+    CARBON_VLOG() << "Lowering " << block_id << "\n";
+    auto* llvm_block = function_lowering.GetBlock(block_id);
+    // Keep the LLVM blocks in lexical order.
+    llvm_block->moveBefore(llvm_function->end());
```

Comment:
Could you avoid this reordering if you did two passes:

1) Create all body blocks.
2) Lower all body blocks.

Then remove the ability for GetBlock to create blocks.

But maybe that'd leave synthetic blocks out of order?

Reply:
Yeah, we'd need some care to put synthetic blocks in the right place, but that'd be possible (we can find the current basic block from the insertion point in the IR builder, and insert the synthetic block after that one). The reason I didn't do that was that this would get in the way of our optimization to avoid creating empty blocks -- we'd either need to move that optimization into step 1, or delete some of the blocks we created in step 1 if we find they end up being empty -- but I think that is a surmountable problem.